rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,540 @@
+
+In the future, we will consider exposing a dedicated attribute for
+block-level API, to make API cleaner. Internally, we can rely on the
+same abstraction for primitive and block-level attributes to simplify
+passing attributes from primitive API calls to internal block level
+API calls. Like for primitive level API, those would be opaque, and
+attributes would be set/get using dedicated setters and getters.
+
+
+## Managing jitting overheads and caching support
+
+To implement a single matrix multiplication operation, the block-level
+API user will typically need to rely on multiple kernels.
+In particular:
+- tail handling on M, N and both M/N can require dedicated kernels if
+  M/N are not multiples of the block-size chosen by the user,
+- when blocking on K dimension, extra kernels are required to handle
+  first K-block (beta=0), intermediate K-blocks (beta=1) and last
+  K-block (beta=1, D matrix set, and post-ops specified).
+- when full matrices dimension change frequently. This will lead to
+  changing leading dimension at the brgemm level, requiring new
+  kernels.
+
+On top of that, if the block-level kernels are generated independantly
+for each layer of a neural network, the same kernels might be
+generated multiple times.
+
+Some API design decisions could impact potential reusability of kernels.
+In particular:
+- passing alpha/beta at runtime instead of creation time.
+- support two execute functions (with and without post-op and final
+  data conversion) for the same brgemm object.
+- support runtime leading dimensions.
+
+The proposal here is to stick with what is already tried and used
+internally. If during the experimental stage of the API, users request
+these functionalities, we will revisit the API and add the necessary
+support in the internal implementation. So alpha/beta will be creation
+time parameters, and kernels with and without post-ops will be handled
+with separate objects.  For runtime leading dimension, the internal
+API supports it but this will not be exposed for now. We can either
+change the API when we introduce the feature, or add a placeholder
+parameter if we are confident this feature will gain traction soon.
+
+Also, note that currently, no kernel-level cache exists for block
+level functionalities, but there are plans to add it. In the meantime,
+here are a few ways to mitigate the lack of kernel-level cache:
+- hoist block-level kernel generation for it to happen once at the
+  beggining of the application
+- to handle tails/K-blocking, if generating multiple kernels is too
+  costly, one can copy A/B blocks to a temporary buffer for
+  tails/first K-block, and rely on a single block-level kernel.
+
+
+## Transforms and transpose
+
+Transormation rountines, for example to pack data in a certain layout,
+are typically hard to implement without using intrinsics or assembly.
+To facilitate packing, we will expose an out-of-place transform
+functionality.
+
+## All-in-all
+
+### Interface proposal
```

Comment:
Glad to see the new APIs hide the details of palette management. And I still have a few concerns regarding the future API usage in Graph Compiler.

Currently Graph Compiler calls internal BRGEMM APIs from compiler IR world in both `base pointer + fixed stride` and `array of pointers` favors. The compiler IR only provides very limited grammar, and the internal APIs require complex structs as parameters and need notable low-level control, which is beyond the capability of IR. Hence we built a few runtime wrappers to manage things like parameters construction, kernel reuse and palette management.

Upon Graph Compiler transfering to MLIR, it's no more feasible to maintain a compiler-specific downstream runtime for upstream MLIR, so we tend to remove the runtime wrappers entirely. It seems that runtime wrappers are still inevitable using the new APIs:
- It still takes fairly complex structs as parameters, which are hard to construct by IR. E.g: converting `base pointer + fixed stride` to `array of pointers`, and the construction of `brgemm::desc`
- The APIs are `C++` based, which requires extra handling for name mangling and STL structures, as mentioned by @ZhennanQin before

I think this is also a common issue for invocation from compiler IR world. Is it possible to provide an alternative sets of C APIs taking only plain parameters and executing in a one-line fashion? That would make our lives a lot easier.

For example:
```
uint64_t brgemm_dispatch(dim_t batch, dim_t M, dim_t N, dim_t K, ...); // return handle of JITed kernel

void set_hw_context(uint64_t kernel_handle);
void release_hw_context(uint64_t kernel_handle);

void execute(void * A[], void * B[], ...); // `array of pointers` interface
void execute(void * A, void * B, uint32_t batch_size, uint64_t stride_A, uint64_t stride_B, ...); // helper function for `base pointer + fixed stride` usage, convert to `array of pointers` interface internally
```

Reply:
@huanghaixin008 oneDNN does never provide C++ API without a C API layer underneath it. You can always switch to that if it's a better choice for MLIR.